src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs

```diff
@@ -885,5 +885,20 @@ private static string GetEnvironmentVariablesBlock(IDictionary<string, string> s
         }
 
         private static string GetErrorMessage(int error) => Interop.Kernel32.GetMessage(error);
+
+        /// <devdoc>
+        ///    <para>
+        ///       Gets
+        ///       the friendly name of the process.
+        ///    </para>
+        /// </devdoc>
+        public string ProcessName
+        {
+            get
+            {
+                EnsureState(State.HaveProcessName);
+                return _processName!;
```

Comment:
We can inline caching of this field, and delete `HaveNonExitedId` and changes in `EnsureState()`. Something like:

```c#
        public string ProcessName => _processName ?? (_processName = GetProcessName())!;
```

Reply:
But wouldn't that lose the checks that `EnsureState` performs (like checking the state is not 'exited' and other things)?

Reply:
It doesn't require a new state; `EnsureState(State.HaveNonExitedId)` and null checks can be inlined in this getter without modifying body of EnsureState.

Reply:
Thank you @am11 - sorry for being slow to understand :)  Initially, I was following the pattern use for 'ProcessInfo'.  I'll do these changes now.  Would you be able to help with profiling these changes?

Reply:
For benchmarks, first clone a second copy of repo. Quick way is:

```pwsh
# this is powershell

$ mv runtime runtime-pr
$ git clone runtime-pr runtime-main
$ cd runtime-main
$ git remote add dotnet https://github.com/dotnet/runtime
$ git fetch
$ git reset --hard dotnet/main
```
Then benchmark dotnet:
```powershell
$ dotnet new --install BenchmarkDotNet.Templates::0.13.1
$ dotnet new benchmark --console-app -n ProcessBenchmarks
$ cd ProcessBenchmarks

# modify Benchmarks.cs

$ dotnet run -c Release -- --job Medium --filter '*' --corerun \
    ../runtime_main/artifacts/bin/testhost/net7.0-Windows-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun \
    ../runtime_pr/artifacts/bin/testhost/net7.0-Windows-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun
```
( or `--job Short`, if you want to get the results real quick )

Reply:
> It doesn't require a new state; `EnsureState(State.HaveNonExitedId)` and null checks can be inlined in this getter without modifying body of EnsureState.

I implemented those changes, and have this failing test:
``` xml
      <test name="System.Diagnostics.Tests.ProcessTests.ProcessName_GetNotStarted_ThrowsInvalidOperationException" type="System.Diagnostics.Tests.ProcessTests" method="ProcessName_GetNotStarted_ThrowsInvalidOperationException" time="0.1419104" result="Fail">
        <failure exception-type="Xunit.Sdk.ThrowsException">
          <message><![CDATA[Assert.Throws() Failure\r\nExpected: typeof(System.InvalidOperationException)\r\nActual:   (No exception was thrown)]]></message>
          <stack-trace><![CDATA[   at System.Diagnostics.Tests.ProcessTests.ProcessName_GetNotStarted_ThrowsInvalidOperationException() in C:\git\runtime\src\libraries\System.Diagnostics.Process\tests\ProcessTests.cs:line 1028]]></stack-trace>
        </failure>
      </test>
```

Reply:
Perhaps, something like this will fix the test:
```c#
            get
            {
                if (_processName == null)
                {
                    EnsureState(State.HaveId);
                    EnsureState(State.HaveNonExitedId);

                    _processName = ProcessManager.GetProcessName(_processId, _machineName);

                    if (_processName == null)
                    {
                        throw new InvalidOperationException(SR.NoProcessInfo);
                    }
                }

                return _processName;
            }
```

Reply:
Perfect! Added a new commit and all (local) tests pass.  I won't be able to look at this for a week or so, so please feel free to jump in!  Thanks for your help!